Add support for multiple package-manager experiments

douglasjacobsen opened the pull request:
This merge performs a lot of internal tasks to enable experiments to have multiple package managers attached to them. An example implementation of this is provided in the `spack-pip` package manager, though in the future this could be abstracted into a base class to make it easier to implement for additional use cases.

Generally, more of the ownership is pushed into the package manager class for the software spec and compiler spec definitions / formatting.

Tests for the spack-pip package manager are added as well.

douglasjacobsen committed:
Enhance existing package managers

This commit adds the name of the package manager to software inventory
information, which allows it to be more easily understood when multiple
package managers are used in the same experiment.

Additionally, some python enhancements are made to the spack and pip
package managers. This allows spack to install a python, and tell Ramble
where this python is. The pip package manager is updated to allow
telling it which python to use as well. These changes are to prepare for
a spack-pip package manager, where spack manages the python installation
and pip can then install packages using this installed python.

douglasjacobsen committed:
Add a spack-pip package manager

This commit adds a combined spack + pip package manager that can be used
to construct more complex environments.

linsword13 commented on var/ramble/repos/builtin/package_managers/spack-lightweight/package_manager.py:
```diff
@@ -742,6 +747,13 @@ def __init__(self, shell="bash", dry_run=False):
 
+    def get_spack_python(self):
+        if self.dry_run:
+            return self.bs_python.copy()
```
Does command_runner have a `copy` method?

douglasjacobsen replied:
Ah, I just copied this from the pip package manager. But i guess you used a different class for `bs_python` there.

linsword13 commented on var/ramble/repos/builtin/package_managers/spack-pip/package_manager.py:
```diff
@@ -0,0 +1,273 @@
+class SpackPip(PackageManagerBase):
+    """Chained spack + pip package manager
+
+    This package manager uses spack to install a spack environment, then using the python that is installed within this environment (or whichever one is available in your path) will install pip packages into a python virtual environment. Experiments will load both environments, in the order of spack then pip.
```
Do we want something like checking that a python package spec is included in the spack env definition?

douglasjacobsen replied:
Yeah, I can add some tests once we decide if we like this.

FWIW, you *would* have spack package specs in the environment if you only used the `pkg_spec` name and not the split `spack_pkg_spec` and `pip_pkg_spec` names.

linsword13 replied:
Yeah I like the implementation. Do you plan to add tests to the current PR?

For my original question, I am more curious if we want to always ensure that the user actually specifies it such that a python would indeed be installed by spack.

douglasjacobsen replied:
Ah sure. I thought you were asking about tests. I hadn't really planned on adding many tests... but I'm open to it.

I'm torn on the specification of packages. I can see both sides of it, and I think it might not be super easy to implement that they require a prefix. But I can look into it as well.

douglasjacobsen replied:
OK, this should be resolved now.

douglasjacobsen committed:
Add package manager controls for unprefixed specs

This commit updates the pacakge manager base class, and software
environments classes to allow control over whether a package manager
allows unprefixed specs or not.

This is to help support the case where multiple package managers are
used in the same experiment. This can prevent an unprefixed spec from
acceidentally being applied to both package managers.

douglasjacobsen committed:
Migrate spec concretization to package managers

This commit moves the concretization / extracton of package and compiler
specs from the workspace class to the package manager classes.

Additionally, the logic on when the inject a prefix is added to the
package manager.

douglasjacobsen committed:
Fix variant issues with existing app definitions

douglasjacobsen committed:
Add tests for spack-pip

These also test multiple package manager functionality.

douglasjacobsen committed:
Update variant merge logic, and push into mixin class

This commit changes the logic for merging variants, to allow a more
scoped approach to reasoning about which variants should be used.

This logic is added into the mixin class, so all objects will have the
same functionality.

Updates are made to all object types to support this.

michaelmckinsey1 reviewed (approved):
I have tested the new `spack-pip` package manager in a use case that successfully installs the [Caliper spack package](https://github.com/spack/spack-packages/blob/develop/repos/spack_repo/builtin/packages/caliper/package.py) with python bindings, and correctly links against a python application built with pip (including its dependencies). This use case is also described in https://github.com/GoogleCloudPlatform/ramble/pull/1239.

I want to note that the importance of the `spack-pip` manager is that a python spack package no longer has to specify all, or any, of it's python dependencies as spack packages, and users can use the existing pip packages, which can be very convenient.

linsword13 commented on var/ramble/repos/builtin/base_classes/object_mixin/base_class.py:
```diff
@@ -45,9 +46,54 @@ def _get_app_inst(self):
+        ApplicationBase = get_base_class("application-base")
+        PackageManagerBase = get_base_class("package-manager-base")
+        WorkflowManagerBase = get_base_class("workflow-manager-base")
+        ModifierBase = get_base_class("modifier-base")
```
Would these somewhat hard-coded base class names be an issue for someone trying to override to use their own base classes? (Unless they reuse the same set of names in their overriding classes.)

Perhaps we can instead using the `origin_type` to figure it out?

douglasjacobsen replied:
I had kind of expected that using the base classes (like this) would be safer than using the `origin_type` but I can definitely swap to the `origin_type` and then if we run into issues convert it back to this.

Basically, it's hard for them to not inherit the base application type but expect it to work. But they could change origin type to something other than application and have it still work.

linsword13 commented on var/ramble/repos/builtin/base_classes/object_mixin/base_class.py:
```diff
@@ -45,9 +46,54 @@ def _get_app_inst(self):
+        provide scoped variant definitions.
+
+        Args:
+            include_modifier (ModifierBase): A single modifier to merge in to resulting set
```
Do you prefer this name or `single_modifier`? I kinda think this is a better name?

douglasjacobsen replied:
I didn't have a strong preference. I had three names I actually tried, and it's nice that I accidentally left this artifact of one of them. I'll change to this one since you like it more.

linsword13 commented on var/ramble/repos/builtin/base_classes/package-manager-base/base_class.py:
```diff
@@ -99,6 +100,10 @@ def runner(self):
 
+    @property
+    def allow_unprefixed_specs(self):
+        return getattr(self, "_allow_unprefixed_specs", True)
```
Perhaps the `_allow_unprefixed_specs` can be set inside `init`, and here it would just be a simple `return self._allow_unprefixed_specs`?

douglasjacobsen replied:
Good suggestion, I'll update this.

linsword13 commented on var/ramble/repos/builtin/base_classes/object_mixin/base_class.py:
```diff
@@ -45,9 +46,54 @@ def _get_app_inst(self):
+        experiment_variants = self.experiment_variants()
+        return app_inst.expander.satisfies(when_key, experiment_variants)
+
+    def experiment_variants(self, single_modifier=None):
```
Curious how much caching can be done here: is it a pure function based on the inputs?

douglasjacobsen replied:
It's kind of a pure function, but also not...

It should be a pure function, but some objects (like the application) call this multiple times while they are getting initialized. So, if we cache the objects too early it makes them wrong.

I added a naive implementation of caching that should pass the test though, so we can see how this works.

douglasjacobsen committed:
Address reviewer feedback

douglasjacobsen committed:
Implement caching of variant sets

douglasjacobsen committed:
Convert ObjectType extraction to use origin_type

This commit updates the logic for extracting the type of an object to
use the origin_type attribute instead of checking if is it an instance
of one of the base classes.

linsword13 reviewed (dismissed):
Great stuff!
